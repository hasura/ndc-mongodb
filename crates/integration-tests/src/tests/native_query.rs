use crate::{connector::Connector, graphql_query, run_connector_query};
use insta::assert_yaml_snapshot;
use ndc_test_helpers::{asc, binop, field, query, query_request, target, variable};

#[tokio::test]
async fn runs_native_query_with_function_representation() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This doesn't affect native queries that don't use the $documents stage.
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query NativeQuery {
                  hello(name: "world")
                }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn runs_native_query_with_collection_representation() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query {
                  title_word_frequencies(
                    where: {count: {_eq: 2}}
                    order_by: {word: Asc}
                    offset: 100
                    limit: 25
                  ) {
                    word
                    count
                  }
                }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn runs_native_query_with_variable_sets() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This means that remote joins are not working in MongoDB 5
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        run_connector_query(
            Connector::SampleMflix,
            query_request()
                .variables([[("count", 1)], [("count", 2)], [("count", 3)]])
                .collection("title_word_frequency")
                .query(
                    query()
                        .predicate(binop("_eq", target!("count"), variable!(count)))
                        .order_by([asc!("_id")])
                        .limit(20)
                        .fields([field!("_id"), field!("count")]),
                )
        )
        .await?
    );
    Ok(())
}
