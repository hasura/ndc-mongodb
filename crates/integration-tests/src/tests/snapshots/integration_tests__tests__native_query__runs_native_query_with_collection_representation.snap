---
source: crates/integration-tests/src/tests/native_query.rs
expression: "graphql_query(r#\"\n                query {\n                  titleWordFrequency(\n                    where: {count: {_eq: 2}}\n                    order_by: {id: Asc}\n                    offset: 100\n                    limit: 25\n                  ) {\n                    id\n                    count\n                  }\n                }\n            \"#).run().await?"
---
data:
  titleWordFrequency:
    - id: Amish
      count: 2
    - id: Amor?
      count: 2
    - id: Anara
      count: 2
    - id: Anarchy
      count: 2
    - id: Anastasia
      count: 2
    - id: Anchorman
      count: 2
    - id: Andre
      count: 2
    - id: Andrei
      count: 2
    - id: Andromeda
      count: 2
    - id: Andr√®
      count: 2
    - id: Angela
      count: 2
    - id: Angelica
      count: 2
    - id: "Angels'"
      count: 2
    - id: "Angels:"
      count: 2
    - id: Angst
      count: 2
    - id: Animation
      count: 2
    - id: Annabelle
      count: 2
    - id: Anonyma
      count: 2
    - id: Anonymous
      count: 2
    - id: Answer
      count: 2
    - id: Ant
      count: 2
    - id: Antarctic
      count: 2
    - id: Antoinette
      count: 2
    - id: Anybody
      count: 2
    - id: Anywhere
      count: 2
errors: ~
